Improve current logging

tcharding opened the pull request:
With this applied the alice logs for btc_eth happy path e2e test look like this:
```
cat api_tests/log/cnd-alice.log
Using config file /tmp/ce24b192-32ea-4ffa-808c-3d9527064279/config.toml
Initializing logging with base level TRACE
[2019-12-24 09:33:46.399][INFO][cnd::seed:75] No seed file found, creating at: /tmp/tmp-9341F4RJkyANdk51/seed.pem
[2019-12-24 09:33:46.497][INFO][cnd::db:66] SQLite database file: /tmp/tmp-9341F4RJkyANdk51/cnd.sqlite
[2019-12-24 09:33:46.500][INFO][cnd:68] Starting with peer_id: QmTH2GwivoaZzdeJ3MLFyxkrjh1uKMpa1j2wSCWQJYVUp8
[2019-12-24 09:33:46.504][DEBUG][cnd::load_swaps:32] loading swaps from database ...
[2019-12-24 09:33:46.505][INFO][cnd:153] Starting HTTP server on 0.0.0.0:45063
[2019-12-24 09:33:50.571][TRACE][cnd::http_api::routes::rfc003::handlers::post_swap:194] New request: Request { swap_id: SwapId(9d0361e7-6d71-4b22-a34e-3332791c8223), alpha_ledger: Bitcoin { network: regtest }, beta_ledger: Ethereum { chain_id: ChainId(17) }, alpha_asset: Amount(100000000 satoshi), beta_asset: EtherQuantity(10000000000000000000), hash_function: Sha256, alpha_ledger_refund_identity: PublicKey(PublicKey { compressed: true, key: PublicKey(f1f21e83df364117a1eeccb46847b38c34bbcb2d439ff1fd1ac3a74aad9d09494b4c05cee406622d1966ab4f4d534e24e49dc49ce8de317fdfc248583e488d51) }), beta_ledger_redeem_identity: 0x69a9eb16cc8d89b71ed547a2e781b246d13983f6, alpha_expiry: Timestamp(3485372400), beta_expiry: Timestamp(3485336400), secret_hash: SecretHash(287dbd12968a9737d1cd00e642e7dea8ff252c0417b4f9b1cc207c2e33161c22) }
[2019-12-24 09:33:50.581][TRACE][cnd::network:402] Making swap request to QmTGN3uRGDnBUv4ayP4XMG9WFgoQGLetdnAcg4kxZ4zhqg@/ip4/127.0.0.1/tcp/43129: swap_id: 9d0361e7-6d71-4b22-a34e-3332791c8223
[2019-12-24 09:33:50.589][DEBUG][libp2p_comit::behaviour:181] connected to QmTGN3uRGDnBUv4ayP4XMG9WFgoQGLetdnAcg4kxZ4zhqg at Dialer { address: "/ip4/127.0.0.1/tcp/43129" }
[2019-12-24 09:33:50.592][TRACE][libp2p_comit::handler:228] transition from WaitingSend to WaitingAnswer
[2019-12-24 09:33:50.592][TRACE][libp2p_comit::handler:228] transition from WaitingAnswer to WaitingAnswer
[2019-12-24 09:33:51.650][TRACE][libp2p_comit::handler:228] transition from WaitingAnswer to WaitingClose
[2019-12-24 09:33:51.807][TRACE][cnd::swap_protocols::init_swap:30] initialising accepted swap: 9d0361e7-6d71-4b22-a34e-3332791c8223
[2019-12-24 09:33:51.808][TRACE][cnd::swap_protocols::rfc003::state_machine:346] creating swap: 9d0361e7-6d71-4b22-a34e-3332791c8223
[2019-12-24 09:33:52.660][TRACE][cnd::http_api::routes::rfc003::handlers::action:142] action->fund
[2019-12-24 09:33:52.819][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 1e4fef04de532283981072a1255eed81e42d9a67022c22e4cef5d4d3f458c3bd
[2019-12-24 09:33:53.829][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 3ad0294d3e7f7c48bda9f25817c33b5a47619d39ec876db445ed4f75985e399c
[2019-12-24 09:33:54.839][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 0e6e33c8893b24576946f26a60e6db1db772d83cc26f9e34f85efdfb22d18783
[2019-12-24 09:33:54.842][DEBUG][cnd::swap_protocols::rfc003::state_machine:380] Transitioning to AlphaDeployed
[2019-12-24 09:33:54.842][DEBUG][cnd::swap_protocols::rfc003::state_machine:397] Transitioning to AlphaFunded
[2019-12-24 09:33:55.852][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 2fd7395f8a3d56319ab9e0040c1049935dc5851c70dd8edab7bbe0e6c27ccd95
[2019-12-24 09:33:55.866][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 2fd7395f8a3d56319ab9e0040c1049935dc5851c70dd8edab7bbe0e6c27ccd95
[2019-12-24 09:33:55.871][TRACE][cnd::btsieve::ethereum::web3_connector:65] Fetched block from web3: 0xb19a…76ef
[2019-12-24 09:33:56.882][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 3f775b96add5644360d9fa767839bd7d90b9999dc770ffa6318bea64facae8a5
[2019-12-24 09:33:56.893][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 3f775b96add5644360d9fa767839bd7d90b9999dc770ffa6318bea64facae8a5
[2019-12-24 09:33:56.895][DEBUG][cnd::swap_protocols::rfc003::state_machine:451] Transitioning to AlphaFundedBetaDeployed
[2019-12-24 09:33:56.897][DEBUG][cnd::swap_protocols::rfc003::state_machine:542] Transitioning to BothFunded
[2019-12-24 09:33:57.748][TRACE][cnd::http_api::routes::rfc003::handlers::action:146] action->redeem
[2019-12-24 09:33:57.920][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 55dbd71e810ed12facbb7fc2e99057c219c5b166758ee323e1798032dca10ea8
[2019-12-24 09:33:57.923][TRACE][cnd::btsieve::ethereum::web3_connector:65] Fetched block from web3: 0x0ace…ab3d
[2019-12-24 09:33:57.929][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 55dbd71e810ed12facbb7fc2e99057c219c5b166758ee323e1798032dca10ea8
[2019-12-24 09:33:57.929][TRACE][cnd::btsieve::ethereum::web3_connector:65] Fetched block from web3: 0x0ace…ab3d
[2019-12-24 09:33:58.947][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 66ea6a7bea58cdaec26bc18ed1257e17991dc7d8b6307072529deeada1a6d20e
[2019-12-24 09:33:58.948][DEBUG][cnd::swap_protocols::rfc003::state_machine:569] Transitioning to AlphaFundedBetaRedeemed
[2019-12-24 09:33:58.958][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 66ea6a7bea58cdaec26bc18ed1257e17991dc7d8b6307072529deeada1a6d20e
[2019-12-24 09:33:59.968][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 30e2b33a265c64c58d42efe298aa8d1d43e43b4285027a6dbbb016ee305f5419
[2019-12-24 09:33:59.985][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 30e2b33a265c64c58d42efe298aa8d1d43e43b4285027a6dbbb016ee305f5419
[2019-12-24 09:34:01.006][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 09161bf7f72e2392977b61ac35e0a12cea492953d6e7d681b2916fa239815a6e
[2019-12-24 09:34:01.010][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 09161bf7f72e2392977b61ac35e0a12cea492953d6e7d681b2916fa239815a6e
[2019-12-24 09:34:02.027][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 1afdb9c100e9aefed21ca9f694fb75a76daddc2d17c077336f98c36d48f24b1e
[2019-12-24 09:34:02.035][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 1afdb9c100e9aefed21ca9f694fb75a76daddc2d17c077336f98c36d48f24b1e
[2019-12-24 09:34:03.050][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 5db2275066a2958447e77d9b36e8ae27c41f7026bb70f571e78d8340f0139a2b
[2019-12-24 09:34:03.054][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 5db2275066a2958447e77d9b36e8ae27c41f7026bb70f571e78d8340f0139a2b
[2019-12-24 09:34:04.071][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 6c1c68bdb79af428a56c3fd7b249e77bc4c3dcf51c48700b17f532579d701f68
[2019-12-24 09:34:04.074][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 6c1c68bdb79af428a56c3fd7b249e77bc4c3dcf51c48700b17f532579d701f68
[2019-12-24 09:34:05.101][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 0fe52c25b0026c7cbbccfd0abc10acadad4c4074ea7b760549011c26d60773b9
[2019-12-24 09:34:05.105][TRACE][cnd::btsieve::bitcoin::bitcoind_connector:78] Fetched block from bitcoind: 0fe52c25b0026c7cbbccfd0abc10acadad4c4074ea7b760549011c26d60773b9
[2019-12-24 09:34:05.120][DEBUG][cnd::swap_protocols::rfc003::state_machine:778] Transitioning to Final
[2019-12-24 09:34:05.120][INFO][cnd::swap_protocols::rfc003::state_machine:362] Swap 9d0361e7-6d71-4b22-a34e-3332791c8223 finished with BothRedeemed
```

thomaseizinger commented on cnd/src/swap_protocols/rfc003/messages.rs:
```diff
@@ -48,6 +58,12 @@ pub struct Decline {
 
+impl Scribe for Decline {
+    fn scribe(&self) -> String {
+        format!("Declining swap: {}", self.swap_id.scribe())
```
This is something that I'd rather have inline with the actual processing logic. IMO, traits like `Display` etc should not assume how a struct is used because they cannot possibly know.

tcharding replied:
The `Scribe` (for want of a better name) trait is _only_ used for logging so it seems reasonable to assume how it is used?

thomaseizinger commented on cnd/src/swap_protocols/swap_id.rs:
```diff
@@ -35,3 +36,10 @@ impl fmt::Display for SwapId {
+impl Scribe for SwapId {
+    fn scribe(&self) -> String {
+        let s = format!("{}", self);
+        s[..8].to_string() // Arbitrary number of human readable digits.
```
:mega: We should always display the full swap id. Being a UUID, it is already human readable.

tcharding replied:
Surely we do not need to log 128 bits just to show a unique swap - its way too verbose.

thomaseizinger replied:
But it removes the ability to take the swap-id from the outside and do a search in the logs.

tcharding replied:
What do you mean, if we use the first 12 characters of the swap_id then one can still grep for them.  This is inline with how we use `--abbrev` in the kernel to pass around commit hashes as only 12 characters in email.

thomaseizinger replied:
To me, it seems like this would be equivalent of `git` would only emit 12 char commit hashes in every log and if someone passes you a full commit hash, you need to first abbreviate it if you want to search `git log` for it.

I am not a big fan of this. It is something you have to teach someone once they are surprised that they cannot just take a swap id and grep a log for it.

tcharding replied:
Ok, I have removed `Scribe` and implemented `Display` on things that we want to log.  Of note the implementation for `Request` is super minimal, adding the ledgers and assets is going to be a punish because of all the traitbounds `<AL, BL, AA, BA>`.  This _should_ be resolved if/when we move to tracing since these can be a part of the context.

thomaseizinger commented on cnd/tests/bitcoin_helper/connector_mock.rs:
```diff
@@ -34,15 +34,15 @@ impl BitcoinConnectorMock {
 
 impl LatestBlock for BitcoinConnectorMock {
-    type Error = ();
+    type Error = String;
```
This is Rust, let's use types :)

tcharding replied:
If this builds I'm happy, I was just keeping the compiler quiet :)

luckysori replied:
Dealt with this in https://github.com/comit-network/comit-rs/pull/1849.

tcharding commented:
I believe the only controversial thing in this PR is the addition of the `Sribe` trait and its usage.  @D4nte and @thomaseizinger can you please give a yes/no to this - I can modify it if either of you thinks its no good.  Thanks.

thomaseizinger commented:
> I believe the only controversial thing in this PR is the addition of the `Sribe` trait and its usage. @D4nte and @thomaseizinger can you please give a yes/no to this - I can modify it if either of you thinks its no good. Thanks.

I am leaning towards a _no_ for the reasons stated in my previous review:

> I am a bit unsure about the Scribe trait though, its usage seems a bit inconsistent:
> 
> - sometimes we refer to the Debug output
> - sometimes you add custom information in front of it
> 
> If we don't like some of the Debug or Display implementations of types we want to print, a less invasive and more modular approach would be to define specific wrapper types for pretty-printing, i.e.: our own Block type that implements Display the way we want. We could then use it like this:
> 
> `log::info("Here is a block: {}", PrettyBlock(my_block))`
> 
> Maybe this newtype could even be generic: `Pretty<T>`.

D4nte commented:
It looks like @thomaseizinger is proposing a new approach to `Scribe` with a new type instead. I'd be keen to sit down and that with him and you to understand the different approaches and make an informed decision.

tcharding commented:
> It looks like @thomaseizinger is proposing a new approach to Scribe with a new type instead. I'd be keen to sit down and that with him and you to understand the different approaches and make an informed decision.

Cool, lets do that.

tcharding commented:
@thomaseizinger I don't fully see the benefit of your suggestion compared to adding a trait.  Will let you explain in person, thanks.

thomaseizinger commented:
In a nutshell:
There are already traits for rendering structs: Display and Debug. I'd prefer sticking to std-traits for defining our behaviour where it makes sense. It allows for broader interop.

To have a different Display/Debug for a given type, I'd therefore define new types that have the Display/Debug impl you want.

Wanting to change the Display impl for a given type in a particular _context_ is IMO a smell that you actually want a different type.

D4nte commented on cnd/src/btsieve/bitcoin/bitcoind_connector.rs:
```diff
@@ -46,15 +46,9 @@ impl LatestBlock for BitcoindConnector {
+            .map_err(Self::Error::Reqwest)
             .and_then(move |mut response| {
-                response.json::<ChainInfo>().map_err(|e| {
-                    log::error!("Error when deserialising the response from bitcoind");
```
Why removing this error?

tcharding replied:
Because the error is propagated back up the call chain.  Following the adage 'either return the error or log it'.  Happy to be corrected, ether on the truth of this adage or on this particular case.

D4nte commented on cnd/src/config/settings.rs:
```diff
@@ -95,7 +95,7 @@ pub enum AllowedOrigins {
 #[derivative(Default)]
 pub struct Logging {
-    #[derivative(Default(value = "LevelFilter::Debug"))]
+    #[derivative(Default(value = "LevelFilter::Trace"))]
```
I think `Debug` or even `Info` is fine here. We just need to put trace in our e2e config files.

tcharding replied:
Set to `Info`, thanks.

D4nte commented on cnd/src/http_api/routes/rfc003/handlers/action.rs:
```diff
@@ -130,10 +134,22 @@ pub async fn handle_action<
-            Action::Redeem(action) => action.into_response_payload(query_params),
-            Action::Refund(action) => action.into_response_payload(query_params),
+            Action::Deploy(action) => {
+                log::trace!("action->deploy");
```
This is a cryptic log line. why the arrow?

```suggestion
                log::trace!("received deploy action.");
```

tcharding replied:
Fixed as suggested, thanks.

tcharding commented on cnd/src/swap_protocols/rfc003/state_machine.rs:
```diff
@@ -0,0 +1,828 @@
+// The state_machine_future derive generates quite complex code...
```
Thanks, done.

thomaseizinger commented:
@tcharding I rebased your branch with the following steps:

1. Find the commit this branch was branched off master: git merge-base master logging -> 26557074d0688a0777b29a447bc5e3eb270e118a
2. Get the diff from Github: https://github.com/comit-network/comit-rs/pull/1786.diff
3. Checkout a new branch locally: `git checkout -b logging-rebased 26557074d0688a0777b29a447bc5e3eb270e118a`
4. Make a commit
5. Rebase onto master: git rebase master
6. Force push to this branch

I'll implement the changes we were talking about on Friday :)

thomaseizinger commented:
@tcharding If you agree with the changes, feel free to remove the `no-mergify` label and get it in :)

tcharding commented:
Removed the `Display` impl blocks.  Added `BlockHash` for logging the block hash of bitcoin blocks and ethereum blocks as discussed with @thomaseizinger.

thomaseizinger commented:
@tcharding Rebased to reflect latest changes in master.
